### PR TITLE
#228 user data takeaway and deletion

### LINKED ISSUE #228: 💜 - User data takeaway & deletion
## Summary

In order to respect user's wishes, it would be nice to provide a way for users to get a copy of all data Miunie has about them in a JSON format.

Additionally, the user should have an option to make Miunie delete all of their data.

## Interested? 😊 

If you'd like to help us by solving this issue, just say so in the comments. ❤️ 
And don't shy away from asking any kind of question you might have.

Should you have any questions at all, feel free to contact either me (@petrspelos) or @DraxCodes on our Discord server:

[Discord](https://discord.gg/cGhEZuk)

_And of course, thank you so much for contributing! 🎉 It's because of amazing people like yourself that Miunie exists._

### PR DESCRIPTION
**Pull request by 1n5an1ty**
## Related Issue

Fixes #228 

## Changes
Fix naming bug in EmbedConstructor
Create PrivacyCommand
Create Miunie.Json project and Implement IJsonParser
Add Direct File Message

## Expected Feedback
Overall code flow and general 5an1ty checks

### CODE REVIEW

**Comment by petrspelos** [comment]
@1n5an1ty, if this is ready for a review, feel free to unmark it as a Draft. :slightly_smiling_face:

**Comment by 1n5an1ty** [comment]
I still have some work to do on this.

**Comment by 1n5an1ty** [comment]
Team, I am still working on this. After implementing the 'remove' user data feature. I noticed that some of the commands (time service) don't check for an existing user and thus look as if the command just didn't do anything/work, while others (profile service) will check for existing and if not found create the user.  Research the best approach to resolve these findings. I think I have a good solution for this, but if you have input on this matter... I am always willing to listen.

**Comment by petrspelos** [comment]
@1n5an1ty, the `MiunieUserProvider` used by the Discord layer to construct a `MiunieUser` ensures the existence of a user account.

So I'm fairly confident no checks for user existence are needed. 😊

**Comment by 1n5an1ty** [comment]
I made some minor changes to this and submitted them for review. I am sure I missed something... just let me know when/if you find something.

File: src/Miunie.Core/Json/IJsonParser.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with Miunie. If not, see <https://www.gnu.org/licenses/>.
+
+namespace Miunie.Core.Json
+{
+    public interface IJsonParser
```
Comment (petrspelos) [inline src/Miunie.Core/Json/IJsonParser.cs]: I think a more standard name would be `IJsonConverter`. :thinking:

File: src/Miunie.Core/Json/IJsonParser.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        string ConvertToJson<T>(T obj);
```
Comment (petrspelos) [inline src/Miunie.Core/Json/IJsonParser.cs]: And for this I'd say `Serialize<T>`. Basically the same as both Newtonsoft JSON and System.Text.Json.

It seems to be a standard so why not use it. :blush:

File: src/MiunieWindows.sln [hunk 3]
```diff
@@ -21,7 +21,9 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Miunie.LiteDbStorage", "Miu
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Miunie.WindowsApp", "Miunie.WindowsApp\Miunie.WindowsApp.csproj", "{1850BE21-FA48-4E67-8649-A39BAE6947A1}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Miunie.Avalonia", "Miunie.Avalonia\Miunie.Avalonia.csproj", "{FEC99E1A-1CB3-4FEE-886E-12D7AA132564}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Miunie.Avalonia", "Miunie.Avalonia\Miunie.Avalonia.csproj", "{FEC99E1A-1CB3-4FEE-886E-12D7AA132564}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Miunie.Json", "Miunie.Json\Miunie.Json.csproj", "{E5E3F7F7-0EFA-42F9-96F1-3F793BCCBB6E}"
```
Comment (petrspelos) [inline src/MiunieWindows.sln]: Don't forget to make these solution changes in `Miunie.sln` as well.

File: src/Miunie.Json/Miunie.Json.csproj [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <CodeAnalysisRuleSet>../.ruleset</CodeAnalysisRuleSet>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
+    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
```
Comment (petrspelos) [inline src/Miunie.Json/Miunie.Json.csproj]: The StyleCop analyzer reference is done globally through our `Directory.Build.props` so no need to reference it.

File: src/Miunie.Json/Miunie.Json.csproj [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <CodeAnalysisRuleSet>../.ruleset</CodeAnalysisRuleSet>
```
Comment (petrspelos) [inline src/Miunie.Json/Miunie.Json.csproj]: The ruleset reference is done globally through our `Directory.Build.props` so no need to reference it.

File: src/Miunie.Json/Miunie.Json.csproj [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  <ItemGroup>
+    <ProjectReference Include="..\Miunie.Core\Miunie.Core.csproj" />
+  </ItemGroup>
+
+
+  <ItemGroup>
+    <AdditionalFiles Include="..\stylecop.json">
+      <Link>stylecop.json</Link>
+    </AdditionalFiles>
+  </ItemGroup>
```
Comment (petrspelos) [inline src/Miunie.Json/Miunie.Json.csproj]: The StyleCop json reference is done globally through our `Directory.Build.props` so no need to reference it.

File: src/Miunie.InversionOfControl/Miunie.InversionOfControl.csproj [hunk 7]
```diff
@@ -4,10 +4,24 @@
...
-    <ProjectReference Include="../Miunie.LiteDbStorage/Miunie.LiteDbStorage.csproj" />
+    <ProjectReference Include="..\Miunie.Discord\Miunie.Discord.csproj" />
+    <ProjectReference Include="..\Miunie.Json\Miunie.Json.csproj" />
+    <ProjectReference Include="..\Miunie.LiteDbStorage\Miunie.LiteDbStorage.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <AdditionalFiles Include="..\stylecop.json">
+      <Link>stylecop.json</Link>
+    </AdditionalFiles>
```
Comment (petrspelos) [inline src/Miunie.InversionOfControl/Miunie.InversionOfControl.csproj]: Same here, our `Directory.Build.props` does this so no need to reference it.

File: src/Miunie.InversionOfControl/Miunie.InversionOfControl.csproj [hunk 8]
```diff
@@ -4,10 +4,24 @@
     <TargetFramework>netstandard2.0</TargetFramework>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
```
Comment (petrspelos) [inline src/Miunie.InversionOfControl/Miunie.InversionOfControl.csproj]: Our `Directory.Build.props` does this so no need to reference it.

File: src/Miunie.InversionOfControl/Miunie.InversionOfControl.csproj [hunk 9]
```diff
@@ -4,10 +4,24 @@
...
+  </ItemGroup>
+
   <ItemGroup>
     <ProjectReference Include="../Miunie.Core/Miunie.Core.csproj" />
-    <ProjectReference Include="../Miunie.Discord/Miunie.Discord.csproj" />
-    <ProjectReference Include="../Miunie.LiteDbStorage/Miunie.LiteDbStorage.csproj" />
+    <ProjectReference Include="..\Miunie.Discord\Miunie.Discord.csproj" />
+    <ProjectReference Include="..\Miunie.Json\Miunie.Json.csproj" />
+    <ProjectReference Include="..\Miunie.LiteDbStorage\Miunie.LiteDbStorage.csproj" />
```
Comment (petrspelos) [inline src/Miunie.InversionOfControl/Miunie.InversionOfControl.csproj]: You're overriding the universal `/` path identifiers with Windows specific `\`.

Let's be nice to our penguin friends. :penguin:

File: src/Miunie.Core/Services/PrivacyService.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            _messages = messages;
+            _users = users;
+            _jsonParser = jsonParser;
+        }
+
+        public async Task OutputUserJsonDataAsync(MiunieUser user)
+        {
+            var userJson = _jsonParser.ConvertToJson(user);
+            await _messages.SendDirectFileMessageAsync(user, userJson, Entities.PhraseKey.USER_PRIVACY_FILE_MESSAGE, user.Name);
```
Comment (petrspelos) [inline src/Miunie.Core/Services/PrivacyService.cs]: That's a lot of redundant data. :thinking: 
The first two parameters are really the only two needed.
The PhraseKey shouldn't be provided, the Discord layer will infer it from the method being called.
And the username is part of the `user` object, so there is no need for it. :blush:

File: src/Miunie.Json/Miunie.Json.csproj (hunk 5 above)
Comment (petrspelos) [reply src/Miunie.Json/Miunie.Json.csproj]: You're still adding the ruleset. :sweat_smile:

File: src/Miunie.Json/Miunie.Json.csproj [hunk 11]
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <CodeAnalysisRuleSet>../.ruleset</CodeAnalysisRuleSet>
```
Comment (petrspelos) [inline src/Miunie.Json/Miunie.Json.csproj]:
```suggestion
```

**Pull request merged**